torchdata/datapipes/iter/transform/bucketbatcher.py

```diff
@@ -253,28 +266,45 @@ def __init__(
         self.datapipe = datapipe
         self.max_token_count = max_token_count
         self.buffer_size = buffer_size
+        self.include_padding = include_padding
 
     def __iter__(self) -> Iterator[DataChunk[T_co]]:
         buffer: List = []
         batch: List = []
         batch_size: int = 0
+        max_length: int = 0
         for d in self.datapipe:
             heapq.heappush(buffer, d)
             if len(buffer) == self.buffer_size:
-                length, token = heapq.heappop(buffer)
-                if batch_size + length > self.max_token_count:
-                    yield DataChunk(batch)
-                    batch = []
-                    batch_size = 0
-                batch.append(token)
-                batch_size += length
+                buffer, batch, batch_size, max_length, data_chunk = self._pop_buffer(
+                    buffer, batch, batch_size, max_length
+                )
+                if data_chunk is not None:
+                    yield data_chunk
         while buffer:
-            length, token = heapq.heappop(buffer)
-            if batch_size + length > self.max_token_count:
-                yield DataChunk(batch)
-                batch = []
-                batch_size = 0
-            batch.append(token)
-            batch_size += length
+            buffer, batch, batch_size, max_length, data_chunk = self._pop_buffer(buffer, batch, batch_size, max_length)
+            if data_chunk is not None:
+                yield data_chunk
         if batch:
             yield DataChunk(batch)
+
+    def _pop_buffer(self, buffer: List, batch: List, batch_size: int, max_length: int):
+        data_chunk_to_yield = None
+        length, token = heapq.heappop(buffer)
+
+        if self.include_padding:
+            max_length = max(length, max_length)
+            new_batch_size = (len(batch) + 1) * max_length
+        else:
+            new_batch_size = batch_size + length
+
+        if new_batch_size > self.max_token_count:
+            data_chunk_to_yield = DataChunk(batch)
+            batch = [token]
+            batch_size = length
+            max_length = length
+        else:
+            batch.append(token)
+            batch_size = new_batch_size
```

Comment:
`max_length` needs to be reset to `length` here.

Reply:
`max_length` should be `max(len(x) for x in batch)` at this point. I think that holds with the code as is. If `self.include_padding` is `False`, it will always stay 0 and unused. If `self.include_padding=True`, `max_length` will be contain the value of `max(length, previous_max_length)` (line 296) when we **do not** yield a batch. Or `max_length` will be equal to `length` if we **do** yield a batch.